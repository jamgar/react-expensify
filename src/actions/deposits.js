import database from '../firebase/firebase'

// Add Deposit
export const addDeposit = (deposit) => ({
  type: 'ADD_DEPOSIT',
  deposit
})

export const startAddDeposit = (depositData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    // Destructure depositData, and set default values if none are passed in
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = depositData
    const deposit = { description, note, amount, createdAt }

    return database.ref(`users/${uid}/deposits`).push(deposit).then((ref) => {
      dispatch(addDeposit({
        id: ref.key,
        ...deposit
      }))
    })
  }
}

// Set Depost
export const setDeposits = (deposits) => ({
  type: 'SET_DEPOSITS',
  deposits
})

export const startSetDeposits = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/deposits`).once('value').then((snapshot) => {
      const deposits = []

      snapshot.forEach((childSnapshot) => {
        deposits.push({
          id: snapshot.key,
          ...childSnapshot.val()
        })
      })

      dispatch(setDeposits(deposits))
    })
  }
}
