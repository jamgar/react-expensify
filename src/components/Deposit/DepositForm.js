import React from 'react'
import moment from 'moment'
import { Link } from 'react-router-dom'
import { SingleDatePicker } from 'react-dates'

const now = moment()

export default class DepositForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.deposit ? props.deposit.description : '',
      note: props.deposit ? props.deposit.note : '',
      amount: props.deposit ? (props.deposit.amount / 100).toString() : '',
      createdAt: props.deposit ? moment(props.deposit.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }
  isInvalid = () => (
    this.state.description === '' || this.state.amount === ''
  )
  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }
  onSubmit = (e) => {
    e.preventDefault()
    const {
      description,
      note,
      amount,
      createdAt
    } = this.state

    if (this.isInvalid()) {
      this.setState(() => ({ error: 'Please enter values for description and amount'}))
    } else {
      this.setState(() => { error: '' })
      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note
      })
    }

  }
  render() {
    const {
      description,
      note,
      amount,
      createdAt,
      calendarFocused,
      error
    } = this.state
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {error && <p className="form__error">{error}</p>}
        <input
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={this.onDateChange}
          focused={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your deposit (optional)"
          value={note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button button--margin-right">Save Deposit</button>
          <Link className="button button--secondary" to="/deposit-dashboard">
            Cancel
          </Link>
        </div>
      </form>
    )
  }
}
