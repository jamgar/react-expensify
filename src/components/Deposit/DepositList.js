import React from 'react'
import { connect } from 'react-redux'
import DepositListItem from './DepositListItem'

export const DepositList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Deposit</div>
      <div className="show-for-desktop">Deposit</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.deposits.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Deposits</span>
          </div>
        ) : (
          props.deposits.map((deposit) => (
            <DepositListItem key={deposit.id} {...deposit} />
          ))
        )
      }
    </div>
  </div>
)

const mapStateToProps = (state) => {
  return {
    deposits: state.deposits
  }
}

export default connect(mapStateToProps)(DepositList)
