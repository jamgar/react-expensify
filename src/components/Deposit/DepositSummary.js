import React from 'react'
import { Link } from 'react-router-dom'

const DepositSummary = () => {
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          DepositSummary
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/deposit-create">Add Deposit</Link>
        </div>
      </div>
    </div>
  )
}

export default DepositSummary
