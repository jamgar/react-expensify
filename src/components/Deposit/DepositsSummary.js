import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'
import selectDeposits from '../../selectors/selectors'
import selectDepositsTotal from '../../selectors/selectors-total'

export const DepositsSummary = ({ depositsCount, depositsTotal}) => {
  const depositWord = depositsCount === 1 ? 'deposit' : 'deposits'
  const formatedTotal = numeral(depositsTotal / 100).format('0,00.00')
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{depositsCount}</span> {depositWord} totalling <span>${formatedTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/deposit-create">Add Deposit</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => {
  const visibleDeposits = selectDeposits(state.deposits, state.filters)
  return {
    depositsCount: visibleDeposits.length,
    depositsTotal: selectDepositsTotal(visibleDeposits)
  }
}
export default connect(mapStateToProps)(DepositsSummary)
