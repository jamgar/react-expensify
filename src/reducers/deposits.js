// Deposit Reducer
const depositsReducerDefaultState = []

export default (state = depositsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_DEPOSIT':
      return [
        ...state,
        action.deposit
      ]
      break;
    case 'SET_DEPOSITS':
      return action.deposits
      break;
    case 'EDIT_DEPOSIT':
      return state.map((deposit) => {
        if (deposit.id === action.id) {
          return {
            ...deposit,
            ...action.updates
          }
        } else {
          return deposit
        }
      })
      break;
    case 'REMOVE_DEPOSIT':
      return state.filter(({ id }) => id !== action.id)
      break;
    default:
      return state
  }
}
