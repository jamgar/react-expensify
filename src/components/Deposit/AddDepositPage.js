import React from 'react'
import { connect } from 'react-redux'
import DepositForm from './DepositForm'
import { startAddDeposit } from '../../actions/deposits'

export class AddDepositPage extends React.Component {
  onSubmit = (deposit) => {
    this.props.startAddDeposit(deposit)
    this.props.history.push('/deposit-dashboard')
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Deposit</h1>
          </div>
        </div>
        <div className="content-container">
          <DepositForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddDeposit: (deposit) => dispatch(startAddDeposit(deposit))
})

export default connect(undefined, mapDispatchToProps)(AddDepositPage)
