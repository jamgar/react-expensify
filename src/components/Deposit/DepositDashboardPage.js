import React from 'react'
import DepositsSummary from './DepositsSummary'
import DepositList from './DepositList'
import DepositListFilters from './DepositListFilters'

const DepositDashboardPage = () => (
  <div>
    <DepositsSummary />
    <DepositListFilters />
    <DepositList />
  </div>
)

export default DepositDashboardPage
