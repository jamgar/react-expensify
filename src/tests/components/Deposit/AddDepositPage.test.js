import React from 'react'
import { shallow } from 'enzyme'
import { AddDepositPage } from '../../../components/Deposit/AddDepositPage'
import deposits from '../../fixtures/deposits'

let startAddDeposit, history, wrapper

beforeEach(() => {
  startAddDeposit = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<AddDepositPage startAddDeposit={startAddDeposit} history={history}/>)
})

test('should render AddDepositPage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle onSubmit', () => {
  wrapper.find('DepositForm').prop('onSubmit')(deposits[1])
  expect(history.push).toHaveBeenLastCalledWith('/deposit-dashboard')
  expect(startAddDeposit).toHaveBeenLastCalledWith(deposits[1])
})
