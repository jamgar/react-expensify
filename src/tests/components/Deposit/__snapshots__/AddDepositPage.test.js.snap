// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddDepositPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Deposit
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <DepositForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
